### DISCUSSION TITLE
draw_image vs. insert vs. composite2?

### DISCUSSION THREAD

**Discussion by TheEssem** [libvips/libvips#2697] [Q&A] [answered]
I was looking through the documentation and found 3 very similar functions for writing/compositing an image on top of another: `vips_draw_image`, `vips_insert`, and `vips_composite2`. What exactly are the main differences between these functions? Is any one faster/more performant than the other for simply putting an image on top of another one? Thanks in advance.

**Comment by jcupitt** [comment, marked as answer]
Hi, yes, they have different uses.

`insert` is very fast and simple -- it just puts one image on top of another. It's the opposite of `crop`.

`composite` does full PDF-style porter-duff composition of a set of layers. It's much slower, of course.

`draw_image` is a paintbox-style function -- it destructively modifies the image you use it on. The whole image needs to be in memory, so you can only use it on small objects, or on disc files which can be directly mapped.

**Reply by TheEssem** [reply]
Thanks for the clarification! :)
